pytest_thawgun/plugin.py

```diff
@@ -47,53 +32,86 @@ def _datetime(self, current_time):
+
+    def set_wall_clock(self, to):
+        if self._freeze_time is not None:
+            self._freeze_time.stop()
+        self._freeze_time = freeze_time(to, tick=self._tick)
+        self._frozen_wall_clock_control = self._freeze_time.start()
+
     async def advance(self, offset):
         assert offset >= 0, "Can't go backwards"
 
-        try:
-            base_time = current_time = self.time()
-            new_time = base_time + offset
-            self.wall_offset = timedelta(seconds=time.time() - self.time())
+        prev_ticking_state = self._tick
+        self._tick = False
 
-            with freeze_time(self._datetime(current_time)) as ft:
-                self.loop.time = lambda: current_time
+        self.set_wall_clock(datetime.now())
+        advance_start_dt = advance_end_dt = datetime.now()
 
-                self.logger.debug('Freeze: %s', datetime.now())
+        self._adjust_loop_clock(self.loop.time())
 
-                await self._drain(base_time)
+        loop_target_time = self.loop.time() + offset
 
-                while self.loop._scheduled:
-                    handle = self.loop._scheduled[0]
+        try:
+            await self._drain(self.loop.time())
 
-                    if handle._when > new_time:
-                        break
+            while self.loop._scheduled and self.loop._scheduled[0]._when <= loop_target_time:
+                handle = self.loop._scheduled[0]
+                prev_drain_time = self.loop.time()
+                this_drain_time = handle._when
 
-                    current_time = handle._when
-                    ft.move_to(self._datetime(current_time))
+                self._frozen_wall_clock_control.tick(this_drain_time - prev_drain_time)
 
-                    self.logger.debug('Advance: %s', self._datetime(current_time))
+                self._adjust_loop_clock(this_drain_time)
+                advance_end_dt = datetime.now()
 
-                    if not handle._cancelled:
-                        handle._run()
-                        handle._callback, handle._args = lambda: None, ()
+                if not handle._cancelled:
+                    handle._run()
+                    handle._callback, handle._args = lambda: None, ()
 
-                    await self._drain(current_time)
+                await self._drain(self.loop.time())
 
-                await self._drain(new_time)
-        finally:
-            self.offset += offset
-            self.loop.time = self.time
+            self._frozen_wall_clock_control.tick(loop_target_time - self.loop.time())
 
-        start, end = (self._datetime(base_time), self._datetime(new_time))
+            self._adjust_loop_clock(loop_target_time)
+            await self._drain(self.loop.time())
+
+            advance_end_dt = datetime.now()
+
+        finally:
+            self._tick = prev_ticking_state
+            self._adjust_loop_clock(loop_target_time)
+            self.set_wall_clock(advance_end_dt)
```

Comment:
`handle._run()` is an internal API of asyncio, which might raise an Exception, and if doesn't do so now, it might in the future. Besides, we'd like to keep the state consistent no matter what when `advance()` finishes.